src/S3/Crypto/S3EncryptionClient.php

```diff
@@ -0,0 +1,381 @@
+                    case HeadersMetadataStrategy::class:
+                        return new HeadersMetadataStrategy();
+                        break;
+                    case InstructionFileMetadataStrategy::class:
+                        return new InstructionFileMetadataStrategy(
+                            $this->client,
+                            $instructionFileSuffix
+                        );
+                        break;
+                    default:
+                        throw new \InvalidArgumentException('Could not match the'
+                            . ' specified string in "MetadataStrategy" to a'
+                            . ' predefined strategy.');
+                }
+            } else {
+                throw new \InvalidArgumentException('The metadata strategy that'
+                    . ' was passed to "MetadataStrategy" was unrecognized.');
+            }
+        } elseif ($instructionFileSuffix) {
+            return new InstructionFileMetadataStrategy(
+                $this->client,
+                $instructionFileSuffix
+            );
+        }
+
+        return null;
+    }
+
+    /**
+     * Encrypts the data in the 'Body' field of $args and promises to upload it
+     * to the specified location on S3.
+     *
+     * @param array $args Arguments for encrypting an object and uploading it
+     *                    to S3 via PutObject.
+     *
+     * The required configuration arguments are as follows:
+     *
+     * - @MaterialsProvider: (MaterialsProvider) Provides Cek, Iv, and Cek
+     *   encrypting/decrypting for encryption metadata.
+     * - @CipherOptions: (array) Cipher options for encrypting data. Only the
+     *   Cipher option is required. Accepts the following:
+     *       - Cipher: (string) cbc|gcm
+     *            See also: AbstractCryptoClient::$supportedCiphers
+     *       - KeySize: (int) 128|192|256
+     *            See also: MaterialsProvider::$supportedKeySizes
+     *       - Aad: (string) Additional authentication data. This option is
+     *            passed directly to OpenSSL when using gcm. It is ignored when
+     *            using cbc.
+     *
+     * The optional configuration arguments are as follows:
+     *
+     * - @MetadataStrategy: (MetadataStrategy|string|null) Strategy for storing
+     *   MetadataEnvelope information. Defaults to using a
+     *   HeadersMetadataStrategy. Can either be a class implementing
+     *   MetadataStrategy, a class name of a predefined strategy, or empty/null
+     *   to default.
+     * - @InstructionFileSuffix: (string|null) Suffix used when writing to an
+     *   instruction file if using an InstructionFileMetadataHandler.
+     *
+     * @return PromiseInterface
+     *
+     * @throws \InvalidArgumentException Thrown when arguments above are not
+     *                                   passed or are passed incorrectly.
+     */
+    public function putObjectAsync(array $args) {
```

Comment:
What's the failed scenario look like? Especially for "getObject"? It will be great if some error scenario (besides invalid arguments) can be tested as well